nixos/modules/security/systemd-harden.nix

```
@@ -0,0 +1,115 @@
+
+{config, lib, ...}: let
+  inherit (lib) types;
+in {
+  options.systemd.services = lib.mkOption {
+    type = types.attrsOf (types.submodule ({
+      name,
+      config,
+      ...
+    }: {
+      options.harden = {
+        enable = lib.mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc ''
+            Basic restrictions for systemd services
+          '';
+        };
+
+        execOnlyNix = lib.mkOption {
+          type = types.bool;
+          default = true;
+          description = lib.mdDoc ''
+            Mark whole system as non-executable with exception for `/nix/store`.
+          '';
+        };
+
+        protectKernel = lib.mkOption {
+          type = types.bool;
+          default = true;
+          description = lib.mdDoc ''
+            Protect kernel internals from being reachable by the service.
+          '';
+        };
+
+        proc = lib.mkOption {
+          type = types.bool;
+          default = true;
+          description = lib.mdDoc ''
+            Restrict view into `/proc` to only allow access for limited subset.
+          '';
+        };
+
+        onlyLocalhost = lib.mkOption {
+          type = types.bool;
+          default = true;
+          description = lib.mdDoc ''
+            Allow listening sockets only on localhost.
+          '';
+        };
+
+        ipSockets = lib.mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc ''
+          Allow using AF_INET and AF_INET6
+          '';
+        };
+      };
+      config.serviceConfig = let
+        cfg = config.harden;
+        inherit (lib) mkDefault;
+      in
+        lib.optionalAttrs cfg.enable ({
+          NoNewPrivileges = mkDefault true;
+          LockPersonality = mkDefault true;
+          RemoveIPC = mkDefault true;
+
+          MemoryDenyWriteExecute = mkDefault true;
+
+          CapabilityBoundingSet = mkDefault [""];
+
+          RestrictNamespaces = mkDefault true;
+          RestrictRealtime = mkDefault true;
+          RestrictSUIDSGID = mkDefault true;
+
+          ProtectHome = mkDefault true;
+          ProtectClock = mkDefault true;
+          ProtectControlGroups = mkDefault true;
+
+          RestrictAddressFamilies = mkDefault (
+            ["AF_UNIX"] ++ lib.optional cfg.ipSockets ["AF_INET" "AF_INET6"]
```

This would result in a nested list. Please also note that we put spaces around braces and brackets in Nixpkgs.

```suggestion
            [ "AF_UNIX" ] ++ lib.optionals cfg.ipSockets [ "AF_INET" "AF_INET6" ]
```